File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not len(lists):
            return None
        while len(lists)>1:
            mergeLists = []
            for i in range(0,len(lists),2):
             
                dummy = ListNode()
                temp = dummy
                while i+1<len(lists) and lists[i] and lists[i+1]:
                    if lists[i].val <= lists[i+1].val:
                        temp.next = lists[i]
                        lists[i] = lists[i].next
                    else:
                        temp.next = lists[i+1]
                        lists[i+1] = lists[i+1].next
                    temp = temp.next
                
                if lists[i]:
                    temp.next=lists[i]
                
                if i+1<len(lists) and lists[i+1]:
                    temp.next=lists[i+1]
              
                mergeLists.append(dummy.next)
            lists = mergeLists
        return lists[0]
    
 

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode dummy;
    struct ListNode* temp = &dummy;
    dummy.next = NULL;

    while (l1 != NULL && l2 != NULL) {
        if (l1->val <= l2->val) {
            temp->next = l1;
            l1 = l1->next;
        } else {
            temp->next = l2;
            l2 = l2->next;
        }
        temp = temp->next;
    }

    if (l1 != NULL) {
        temp->next = l1;
    } else {
        temp->next = l2;
    }

    return dummy.next;
}

struct ListNode* mergeKLists(struct ListNode** lists, int listsSize) {
    if (listsSize == 0) {
        return NULL;
    }

    while (listsSize > 1) {
        int mergedSize = (listsSize + 1) / 2;
        for (int i = 0; i < listsSize / 2; i++) {
            lists[i] = mergeTwoLists(lists[i], lists[listsSize - 1 - i]);
        }
        listsSize = mergedSize;
    }

    return lists[0];
}